pkg/scrape/discovery/aws/ec2.go

```diff
@@ -0,0 +1,330 @@
+// limitations under the License.
+
+package aws
+
+import (
+	"context"
+	"fmt"
+	"github.com/prometheus/prometheus/util/strutil"
+	"github.com/sirupsen/logrus"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+
+	"github.com/pyroscope-io/pyroscope/pkg/scrape/discovery"
+	"github.com/pyroscope-io/pyroscope/pkg/scrape/discovery/refresh"
+	"github.com/pyroscope-io/pyroscope/pkg/scrape/discovery/targetgroup"
+	"github.com/pyroscope-io/pyroscope/pkg/scrape/model"
+)
+
+const (
+	ec2Label                  = model.MetaLabelPrefix + "ec2_"
+	ec2LabelAMI               = ec2Label + "ami"
+	ec2LabelAZ                = ec2Label + "availability_zone"
+	ec2LabelAZID              = ec2Label + "availability_zone_id"
+	ec2LabelArch              = ec2Label + "architecture"
+	ec2LabelIPv6Addresses     = ec2Label + "ipv6_addresses"
+	ec2LabelInstanceID        = ec2Label + "instance_id"
+	ec2LabelInstanceLifecycle = ec2Label + "instance_lifecycle"
+	ec2LabelInstanceState     = ec2Label + "instance_state"
+	ec2LabelInstanceType      = ec2Label + "instance_type"
+	ec2LabelOwnerID           = ec2Label + "owner_id"
+	ec2LabelPlatform          = ec2Label + "platform"
+	ec2LabelPrimarySubnetID   = ec2Label + "primary_subnet_id"
+	ec2LabelPrivateDNS        = ec2Label + "private_dns_name"
+	ec2LabelPrivateIP         = ec2Label + "private_ip"
+	ec2LabelPublicDNS         = ec2Label + "public_dns_name"
+	ec2LabelPublicIP          = ec2Label + "public_ip"
+	ec2LabelSubnetID          = ec2Label + "subnet_id"
+	ec2LabelTag               = ec2Label + "tag_"
+	ec2LabelVPCID             = ec2Label + "vpc_id"
+	ec2LabelSeparator         = ","
+)
+
+// DefaultEC2SDConfig is the default EC2 SD configuration.
+var DefaultEC2SDConfig = EC2SDConfig{
+	Port:            80,
+	RefreshInterval: 60 * time.Second,
+}
+
+func init() {
+	discovery.RegisterConfig(&EC2SDConfig{})
+}
+
+// EC2Filter is the configuration for filtering EC2 instances.
+type EC2Filter struct {
+	Name   string   `yaml:"name"`
+	Values []string `yaml:"values"`
+}
+
+// EC2SDConfig is the configuration for EC2 based service discovery.
+type EC2SDConfig struct {
+	Endpoint        string        `yaml:"endpoint"`
+	Region          string        `yaml:"region"`
+	AccessKey       string        `yaml:"access_key,omitempty"`
+	SecretKey       string        `yaml:"secret_key,omitempty"`
+	Profile         string        `yaml:"profile,omitempty"`
+	RoleARN         string        `yaml:"role_arn,omitempty"`
+	RefreshInterval time.Duration `yaml:"refresh_interval,omitempty"`
+	Port            int           `yaml:"port"`
+	Filters         []*EC2Filter  `yaml:"filters"`
+}
```

Comment:
This is something we may support in future but as of now pyroscope server uses **kebab-case** for yaml:
```suggestion
type EC2SDConfig struct {
	Endpoint        string        `yaml:"endpoint"`
	Region          string        `yaml:"region"`
	AccessKey       string        `yaml:"access-key,omitempty"`
	SecretKey       string        `yaml:"secret-key,omitempty"`
	Profile         string        `yaml:"profile,omitempty"`
	RoleARN         string        `yaml:"role-arn,omitempty"`
	RefreshInterval time.Duration `yaml:"refresh_interval,omitempty"`
	Port            int           `yaml:"port"`
	Filters         []*EC2Filter  `yaml:"filters"`
}
```

Reply:
ack,noted, i guess also `refresh_interval` need to be changed？

Reply:
hi @kolesnikovae, could you please help review this?

Reply:
Awesome work @futurelm! Thank you so much!